### PR TITLE
Special characters should be forbidden in mailbox name

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 1]
```diff
@@ -48,6 +52,10 @@ public static MailboxPath forUser(String username, String mailboxName) {
         return new MailboxPath(MailboxConstants.USER_NAMESPACE, username, mailboxName);
     }
 
+    private static final String INVALID_CHARS = "%*&#";
+    private static final CharMatcher INVALID_CHARS_MATCHER = CharMatcher.anyOf(INVALID_CHARS);
+    public  static final int MAX_MAILBOX_NAME_LENGTH = 200;
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: why? In fact when we are adding such value, I think we should add a reference to the authoritative RFC.

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MailboxPathTest.java [hunk 2]
```diff
@@ -181,6 +188,55 @@ public void hasEmptyNameInHierarchyShouldBeTrueIfPathWithTrailingEmptyNames() {
             .isTrue();
     }
 
+    @Test
+    public void assertAcceptable1() {
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MailboxPathTest.java]: 1?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: > why? In fact when we are adding such value, I think we should add a reference to the authoritative RFC.

It's not about RFC, it's some JPA string field size at the origin of this limit.

Past code, past decision, just refactored. Is it worth getting this debate?

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MailboxPathTest.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/api/src/test/java/org/apache/james/mailbox/model/MailboxPathTest.java]: mistyped.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java (hunk 1 above)
Comment (rouazana) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: no debate, but then just a little comment explaining where this value comes from would be great.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 3]
```diff
@@ -136,7 +144,27 @@ public MailboxPath sanitize(char delimiter) {
...
-    public boolean hasEmptyNameInHierarchy(char pathDelimiter) {
+    public void assertAcceptable(char pathDelimiter) throws MailboxNameException {
+        if (hasEmptyNameInHierarchy(pathDelimiter)) {
+            throw new HasEmptyMailboxNameInHierarchyException(asString());
+        }
+        if (nameContainsForbiddenCharacters()) {
+            throw new MailboxNameException(asString() + " contains one of the forbidden characters " + INVALID_CHARS);
+        }
+        if (isMailboxNameTooLong()) {
+            throw new TooLongMailboxNameException("Mailbox name exceed maximum size of " + MAX_MAILBOX_NAME_LENGTH + " characters");
```
Comment (remk) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: s/exceed/exceeds the

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java [hunk 4]
```diff
@@ -48,6 +52,10 @@ public static MailboxPath forUser(String username, String mailboxName) {
         return new MailboxPath(MailboxConstants.USER_NAMESPACE, username, mailboxName);
     }
 
+    private static final String INVALID_CHARS = "%*&#";
```
Comment (remk) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: From the RFC i have seen only that the use of `%` and `*` are discouraged.
What is the reason we forbid to use '%' and `#` ?

Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxPath.java]: The RFC states:

```
   5)    Two characters, "#" and "&", have meanings by convention, and
         should be avoided except when used in that convention.
```

`#` is a special character identifying fully qualified names as oposed to relatives names. This is the namespace prefix.

Regarding `&` I would bet this is related to modified UTF-7 - it's use would require potentially buggy escaping strategies.

```
   "&" is used to shift to modified BASE64 and "-" to shift back to
   US-ASCII.  There is no implicit shift from BASE64 to US-ASCII, and
   null shifts ("-&" while in BASE64; note that "&-" while in US-ASCII
   means "&") are not permitted.  However, all names start in US-ASCII,
   and MUST end in US-ASCII; that is, a name that ends with a non-ASCII
   ISO-10646 character MUST end with a "-").
```

**Comment by mbaechler** [comment]
Any change it breaks some existing deployments?

**Comment by chibenwa** [comment]
That is only enforced on newly created mailboxes, we should be fine.

I followed same strategy than ..

**Comment by chibenwa** [comment]
:green_apple: before CI restart

**Comment by chibenwa** [comment]
```
[c2fb26415743d159805a28d8c96b80a5c0931871] [ERROR] Failures: 
[c2fb26415743d159805a28d8c96b80a5c0931871] [ERROR]   EventDeadLettersIntegrationTest.groupIdOfFailedEventShouldBeStoredInDeadLetter:240 1 expectation failed.
[c2fb26415743d159805a28d8c96b80a5c0931871] JSON path . doesn't match.
[c2fb26415743d159805a28d8c96b80a5c0931871] Expected: iterable with items ["org.apache.james.webadmin.integration.EventDeadLettersIntegrationTest$RetryEventsListenerGroup"] in any order
[c2fb26415743d159805a28d8c96b80a5c0931871]   Actual: []
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**